test/infrastructure/inmemory/internal/controllers/inmemorymachine_controller.go

```diff
@@ -487,6 +513,54 @@ func (r *InMemoryMachineReconciler) reconcileNormalETCD(ctx context.Context, clu
 	return ctrl.Result{}, nil
 }
 
+type etcdInfo struct {
+	clusterID string
+	leaderID  string
+	members   sets.Set[string]
+}
+
+func (r *InMemoryMachineReconciler) inspectEtcd(ctx context.Context, cloudClient cclient.Client) (etcdInfo, error) {
+	etcdPods := &corev1.PodList{}
+	if err := cloudClient.List(ctx, etcdPods,
+		client.InNamespace(metav1.NamespaceSystem),
+		client.MatchingLabels{
+			"component": "etcd",
+			"tier":      "control-plane"},
+	); err != nil {
+		return etcdInfo{}, errors.Wrap(err, "failed to list etcd members")
+	}
+
+	if len(etcdPods.Items) == 0 {
+		return etcdInfo{}, nil
+	}
+
+	info := etcdInfo{
+		members: sets.New[string](),
+	}
+	var leaderFrom time.Time
+	for _, pod := range etcdPods.Items {
+		info.clusterID = pod.Annotations[cloudv1.EtcdClusterIDAnnotationName]
```

Comment:
I think it cannot happen (as we create the initial Machine first and wait for it), but should we return an error if we find etcd pods with different clusterIDs?

Could just make triaging a lot easier in case it somehow is actually possible

(same in etcd/handler.go)

Reply:
Added, but as far as I remember cluster-ID is not relevant for KCP and cache's resource groups already provide a tenancy boundary that prevents etcd pod/member from two clusters to mix up

Reply:
Yup should be safe already. Just thought it's good to return an error in this case. Just in case things are changing and we hit this issue at some point. Then it's probably great that this issue is simply surfaced vs. us having to debug how some weird effects happened because of that